### DISCUSSION TITLE
Unable to use @web3modal/ethereum due to ERR_REQUIRE_ESM

### DISCUSSION THREAD

**Discussion by mohandast52** [WalletConnect/walletconnect-monorepo#2554] [Developer Support]
Hello team, 

I wanted to inform you that we have a dedicated library called [autonolas/library](https://www.npmjs.com/package/@autonolas/frontend-library) which houses all our common components. Recently, I created a new component called LoginV2 as part of our migration from V1 to V2. The component functions perfectly within our storybook, as shown in the attached screenshot.

However, I am facing an issue when trying to import the component into other libraries, and I'm unable to determine how to resolve it. Below, I have included the configuration I've been using. Any assistance or guidance on this matter would be highly appreciated. Thank you for your support.

tsconfig of common library:
```json
{
  "compilerOptions": {
    "baseUrl": "./",
    "allowSyntheticDefaultImports": true,
    "module": "commonjs",
    "target": "es5",
    "lib": ["ES2015", "es6", "dom", "es2020"],
    "sourceMap": true,
    "allowJs": false,
    "jsx": "react",
    "moduleResolution": "node",
    "rootDir": "src",
    "outDir": "dist",
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "declaration": true,
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "strict": true,
  },
  "include": ["src/**/*", "src/jest-setup.ts"],
  "exclude": [
    "node_modules",
    "build",
    "dist",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "**/*/*.test.ts",
    "examples"
  ]
}
```

.babelrc of the importing repository
```json
{
  "presets": [["next/babel"], ["@babel/preset-env"]],
  "plugins": [
    ["@babel/plugin-transform-modules-commonjs"],
    ["@babel/plugin-transform-runtime"],
    [
      "module-resolver",
      {
        "extensions": [".js", ".jsx"],
        "alias": {
          "util": "./util",
          "common-util": "./common-util",
          "components": "./components",
          "images": "./public/images",
          "store": "./store"
        }
      }
    ],
    [
      "styled-components",
      {
        "ssr": true,
        "displayName": true,
        "preprocess": false
      }
    ],
    ["inline-react-svg"],
    [
      "import",
      { "libraryName": "antd", "libraryDirectory": "lib", "style": true }
    ]
  ]
}
```
<img width="1226" alt="login" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/22061815/46fd625f-5f4a-4b6f-977d-4bc4ee7c103e">

More info: I use `nextJS` and `antd` in the importing library

**Comment by glitch-txs** [comment]
Seems like an issue with bundler configuration more than related to WalletConnect, are you bundling for both commonjs and esm? Could you show the full log?

**Reply by mohandast52** [reply]
Sure, let me share the error log
```
command % yarn dev
yarn run v1.22.19
$ next
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
info  - Loaded env from project-name/.env
project-name/node_modules/@autonolas/frontend-library/dist/components/common/LoginV2/LoginV2.js:28
var ethereum_1 = require("@web3modal/ethereum");
                 ^

Error [ERR_REQUIRE_ESM]: require() of ES Module project-name/node_modules/@web3modal/ethereum/dist/index.es.js from project-name/node_modules/@autonolas/frontend-library/dist/components/common/LoginV2/LoginV2.js not supported.
Instead change the require of index.es.js in project-name/node_modules/@autonolas/frontend-library/dist/components/common/LoginV2/LoginV2.js to a dynamic import() which is available in all CommonJS modules.
    at Object.<anonymous> (project-name/node_modules/@autonolas/frontend-library/dist/components/common/LoginV2/LoginV2.js:28:18)
    at Object.<anonymous> (project-name/node_modules/@autonolas/frontend-library/dist/components/common/LoginV2/index.js:4:17)
    at Object.<anonymous> (project-name/node_modules/@autonolas/frontend-library/dist/components/common/index.js:19:14)
    at Object.<anonymous> (project-name/node_modules/@autonolas/frontend-library/dist/components/index.js:17:14)
    at Object.<anonymous> (project-name/node_modules/@autonolas/frontend-library/dist/index.js:17:14)
    at Object.@autonolas/frontend-library (project-name/.next/server/pages/_app.js:228:18)
    at __webpack_require__ (project-name/.next/server/webpack-runtime.js:33:42)
    at Function.__webpack_require__.t (project-name/.next/server/webpack-runtime.js:58:38) {
  code: 'ERR_REQUIRE_ESM'
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Reply by glitch-txs** [reply]
Mmm yes this is related to the bundler, make sure you have the proper config, I'm not very familiar with babel tbh. I can recommend `tsup` though I just built a lib with it using WalletConnect 😸

**Reply by mohandast52** [reply]
😅 I recently made some changes in the code to switch from CommonJS to support the ESM module. However, this change ended up causing issues in other parts of the code. I also made some adjustments to the Babel configurations, but unfortunately, the issue still persists. I'll have a look at `tsup` and see if it can provide any useful insights or solutions. In the meantime, if you come across any helpful information or suggestions, please let me know. Thank you for your assistance 🙏

**Reply by glitch-txs** [reply]
You should add support for both of them. Regarding tsup, I'm pretty new to building libraries and it was very easy to set up, I also use next.js, I can show you the config I used if you want, if you're not tight to babel. Otherwise you can show me your config I can try helping you out

**Reply by mohandast52** [reply]
Sure, please share your config, that would be helpful 🙌 and I have already added the configs (`tsconfig` of the library and `.babelrc` config of the importing repository) in the description. Please let me know if you require anything else.

**Reply by glitch-txs** [reply]
I'll try to do some research later

Check tsconfig, tsup, and package.json all of them matter

You basically install tsup as dev dep

https://github.com/glitch-txs/w3/blob/main/packages/core/tsup.config.ts

What I haven't figured out yet is the support for app dir, it has a workaround on the next.js side though

**Reply by mohandast52** [reply]
Hello @glitch-txs, I followed your suggestion and tried using `tsup`, but unfortunately, the issue persists 😭. Have you had a chance to look into it and figure out a workaround? I'm also curious as to why web3modal is not compiled to commonJS?

**Reply by glitch-txs** [reply]
Is your project public? Or if you have a minimal repro example would be great too

**Reply by mohandast52** [reply]
yes, it is public.
This is the frontend-library => https://github.com/valory-xyz/autonolas-frontend-library
This will be consuming repository => https://github.com/valory-xyz/autonolas-protocol-frontend (nextJS powered)

**Reply by glitch-txs** [reply]
Do you get the same error with `npm link`?

**Reply by mohandast52** [reply]
yes and I think it's the same because it will be compiled and added to `/dist` folder

**Reply by ahollenbach** [reply]
Hello! We also ran into this issue with the same setup (importing `@walletconnect/modal` into another library), using tsup to generate both cjs and esm exports. We only run into the issue if we include the modal, not the base ethereum-provider.

I noticed that your rollup configs seem to vary slightly in the [ethereum-provider repo](https://github.com/WalletConnect/walletconnect-monorepo/blob/078b69503f612e100506753cc6a1d820307df79a/rollup.config.js#L44) vs in the [@walletconnect/modal repo](https://github.com/WalletConnect/modal/blob/5d792fe57f85944482874e4472c3aff796cd9d72/rollup.config.js#L45).

**Reply by glitch-txs** [reply]
Hi guys, this config worked for me, for tsup I could only make it work for core package using ethereum-provider as well.
For React I could make it work with an old rollup config, you can check both here
https://github.com/glitch-txs/w3/tree/main/packages

The react one is inside the UI folder btw*

**Reply by ahollenbach** [reply]
Thanks @glitch-txs - I can take a look at what tsup is doing differently from rollup and report back.
